account_anglo_saxon_stock_move/model/invoice.py

```diff
@@ -43,87 +43,70 @@ def line_get_convert(self, line, part, date):
 
     @api.multi
     def action_cancel(self):
-        aml_obj = self.env['account.move.line']
-        aml_ids = []
-        for inv in self:
-            if not inv.move_id:
-                continue
-            for aml_brw in inv.move_id.line_id:
-                # Unreconcile all non-receivable and non-payable lines
-                if aml_brw.account_id.reconcile and aml_brw.account_id.type \
-                        not in ('receivable', 'payable'):
-                    aml_ids.append(aml_brw.id)
-        if aml_ids:
-            aml_obj._remove_move_reconcile(aml_ids)
-
+        for inv in self.filtered(lambda inv: inv.move_id):
+            amr_ids = inv.invoice_line.mapped('move_id.aml_ids.reconcile_id')
+            amr_ids.unlink()
         return super(AccountInvoice, self).action_cancel()
 
-    def reconcile_stock_accrual(self, cr, uid, ids, context=None):
-        context = dict(context or {})
-        ids = isinstance(ids, (int, long)) and [ids] or ids
-        aml_obj = self.pool['account.move.line']
-        amr_obj = self.pool['account.move.reconcile']
-        for inv_brw in self.browse(cr, uid, ids, context=context):
-            for ail_brw in inv_brw.invoice_line:
-                ail_brw.refresh()
-                if not ail_brw.move_id:
-                    continue
-                if ail_brw.move_id.product_id != ail_brw.product_id:
+    @api.multi
+    def reconcile_stock_accrual(self):
+        aml_obj = all_aml_ids = self.env['account.move.line']
+        for inv_brw in self:
+            # We just care about lines which have a stock_move.aml_ids related
+            aml_ids = inv_brw.invoice_line.mapped('move_id.aml_ids')
+
+            # In order to keep every single line reconciled we will look for
+            # all the lines related to a purchase/sale order
+            all_aml_ids |= aml_ids.mapped('purchase_id.aml_ids')
+            all_aml_ids |= aml_ids.mapped('sale_id.aml_ids')
+
+            categ_ids = all_aml_ids.filtered(
+                lambda m:
+                m.product_id and
+                not m.product_id.categ_id.property_stock_journal)
+            if categ_ids:
+                raise ValidationError(_(
+                    'The Stock Journal is missing on following '
+                    'product categories: %s' % (', '.join(
+                        categ_ids.mapped('name')))
+                ))
+
+            res = {}
+            # Only stack those that are fully reconciled
+            amr_ids = all_aml_ids.mapped('reconcile_id')
+            amr_ids.unlink()
+
+            # Let's group all the Accrual lines by Purchase/Sale Order, Product
+            # and Account
+            for aml_brw in all_aml_ids.filtered(
+                    lambda l: l.account_id.reconcile):
+                doc_brw = aml_brw.purchase_id or aml_brw.sale_id
+                account_id = aml_brw.account_id.id
+                product_id = aml_brw.product_id
```

Comment:
`account_id = aml_brw.account_id.id` in here I use it group by account_id (integer) 
but here `product_id = aml_brw.product_id` I use it as an object because I want it 
to get advantage of object [here](https://github.com/Vauxoo/addons-vauxoo/pull/1218/files#diff-f6d45b6288783ab3ee76bc9962939e08R90)